File: code/python/scripts/run_pilot.py
"""Banco principal: reguladores de orden 0, 1 y 2 sobre los cuatro escenarios.

PROTOCOLO
---------
  * ingredientes terminales CORRECTOS (S_i = 0 por construccion)
  * 1er y 2o orden IGUALADOS en atenuacion a Nyquist EN TIEMPO DISCRETO
  * anti-windup que actua SOLO cuando el filtro bloquea de verdad
  * coste de evaluacion NEUTRAL: ponderado por la demanda limpia, no por el
    peso que cada regulador decide (si no, cada uno se puntuaria con su propio
    criterio y la comparacion no significaria nada)
  * contrastes pareados por semilla, con correccion de Holm por familia

PREDICCION DECLARADA ANTES DE MIRAR NINGUN NUMERO
-------------------------------------------------
  A (demanda cuasi-estatica) : el 2o orden PIERDE frente al 1er orden. Debe
      perder: sobreoscila donde no hace falta, y su pico resonante amplifica
      ruido donde no hay senal util que recuperar.
  B, C, D (demanda no estacionaria y fuera del sobre de diseno) : el 2o orden
      gana. Es la misma firma que el paper del HBP: ninguna ventaja en el
      regimen facil, ventaja fuera de el.
  COSTE: seguir mejor la demanda cuesta mas. En un banco de juguete la demanda
      no vale lo que cuesta seguirla, asi que el peso congelado sera el mas
      barato. Eso NO invalida nada: senala que el banco definitivo necesita que
      seguir la demanda tenga valor real (p.ej. precio horario de la energia).
"""
from __future__ import annotations

import json
import os
import sys
import time

import numpy as np

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from ghi import experiment as gx
from ghi import plant as gplant
from ghi import regulators as greg
from ghi import stats as gs
from ghi import terminal as gterm
from ghi.mompc import MOMPC


def main(n_seeds: int = 10, w0: float = 0.9, zeta: float = 0.5,
         theta: float = 0.5, grid_points: int = 21) -> dict:
    t0 = time.time()
    prob = gplant.problem_conflict()
    prob, term = gterm.design(prob)
    M = MOMPC(prob, term)

    tau, g = greg.match_first_order(w0, zeta, 1.0)
    print("=" * 78)
    print("BANCO PRINCIPAL DEL GHI")
    print("=" * 78)
    print(f"planta      : {prob.name}, N = {prob.N}")
    print(f"terminal    : Kf = {np.round(term.Kf, 4).tolist()}, "
          f"Omega con {term.H.shape[0]} filas")
    print(f"igualacion  : w0={w0} zeta={zeta} -> |H| en Nyquist = {g:.6f}, tau = {tau:.4f}")
    print(f"              (EN DISCRETO; igualar en continuo daba un desajuste de 1.69x)")
    print(f"anti-windup : theta = {theta} (fraccion de velocidad conservada al bloquear)")
    print(f"semillas    : {n_seeds}\n")

    regs = {
        "orden-0": lambda: greg.Order0(),
        "orden-1": lambda: greg.Order1(tau),
        "orden-2": lambda: greg.Order2(w0, zeta, theta),
        "congelado": lambda: greg.Frozen(0.5),
    }
    bank = gx.run_bank(M, regs, gx.default_scenarios(), range(n_seeds),
                       grid_points=grid_points)

    table = gs.contrast_table(
        bank, [("orden-2", "orden-1"), ("orden-2", "orden-0"), ("orden-1", "orden-0")])
    print("=" * 78)
    print("CONTRASTES PAREADOS POR SEMILLA")
    gs.print_contrasts(table)

    print(f"\ntiempo: {time.time() - t0:.1f}s")
    return {"bank": bank,
            "contrasts": {m: [vars(p) for p in fam] for m, fam in table.items()},
            "config": {"w0": w0, "zeta": zeta, "theta": theta, "tau": tau,
                       "g_nyquist": g, "n_seeds": n_seeds}}


if __name__ == "__main__":
    res = main()
    here = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
    dst = os.path.join(here, "results")
    os.makedirs(dst, exist_ok=True)
    with open(os.path.join(dst, "pilot.json"), "w") as fh:
        json.dump(res, fh, indent=1, default=float)
    print(f"escrito {os.path.join(dst, 'pilot.json')}")

File: code/matlab/scripts/run_pilot.m
function out = run_pilot(nSeeds, w0, zeta, theta, gridPoints)
%RUN_PILOT Banco principal del GHI en MATLAB.
%
%   Los numeros de referencia del proyecto salen de la implementacion Python
%   (mas rapida: usa OSQP con setup cacheado). El papel de esta version es
%   confirmar que la MISMA formulacion, resuelta con quadprog, da el mismo
%   resultado cualitativo y los mismos signos en los contrastes.
%
%   Uso:  out = run_pilot();          % 10 semillas
%         out = run_pilot(5);         % mas rapido

if nargin < 1 || isempty(nSeeds),     nSeeds = 10;   end
if nargin < 2 || isempty(w0),         w0 = 0.9;      end
if nargin < 3 || isempty(zeta),       zeta = 0.5;    end
if nargin < 4 || isempty(theta),      theta = 0.5;   end
if nargin < 5 || isempty(gridPoints), gridPoints = 21; end

here = fileparts(mfilename('fullpath'));
addpath(fileparts(here));
t0 = tic;

prob = ghi.problemConflict();
[prob, term] = ghi.designTerminal(prob);
M = ghi.MOMPC(prob, term);
[tau, g] = ghi.matchFirstOrder(w0, zeta, 1.0);

fprintf('==============================================================================\n');
fprintf('BANCO PRINCIPAL DEL GHI -- MATLAB\n');
fprintf('==============================================================================\n');
fprintf('planta      : %s, N = %d\n', prob.name, prob.N);
fprintf('terminal    : Kf = [%.4f %.4f], Omega con %d filas\n', ...
        term.Kf(1), term.Kf(2), size(term.H,1));
fprintf('igualacion  : w0=%.2f zeta=%.2f -> |H| en Nyquist = %.6f, tau = %.4f\n', ...
        w0, zeta, g, tau);
fprintf('              (EN DISCRETO; igualar en continuo daba un desajuste de 1.69x)\n');
fprintf('anti-windup : theta = %.2f\n', theta);
fprintf('semillas    : %d\n\n', nSeeds);

make = {@() ghi.Order0(), @() ghi.Order1(tau), ...
        @() ghi.Order2(w0, zeta, theta), @() ghi.Frozen(0.5)};
rname = {'orden-0','orden-1','orden-2','congelado'};

sc = ghi.defaultScenarios();
snames = fieldnames(sc);
out = struct();

for si = 1:numel(snames)
    s = sc.(snames{si});
    fprintf('### %s\n', s.name);
    fprintf('  %-12s%18s%17s%10s\n', 'regulador', 'err_demanda', 'coste', 'bloqueos');
    for ri = 1:numel(make)
        E = []; C = []; B = [];
        for seed = 0:nSeeds-1
            tr = ghi.closedLoop(M, make{ri}(), s, seed, gridPoints);
            if isempty(tr), continue, end
            k = (s.warmup+1):numel(tr.alpha);
            E(end+1) = sqrt(mean((tr.alpha(k) - tr.clean(k)).^2)); %#ok<AGROW>
            C(end+1) = sum(tr.stage(k));                            %#ok<AGROW>
            B(end+1) = sum(tr.blocked(k));                          %#ok<AGROW>
        end
        out.(snames{si}).(matlab.lang.makeValidName(rname{ri})) = ...
            struct('err', E, 'coste', C, 'bloq', B);
        fprintf('  %-12s%9.4f+-%-7.4f%10.2f+-%-5.2f%10.1f\n', ...
                rname{ri}, mean(E), std(E), mean(C), std(C), mean(B));
    end
    fprintf('\n');
end

fprintf('==============================================================================\n');
fprintf('CONTRASTES PAREADOS POR SEMILLA (delta > 0 => gana el primero)\n');
cmp = {'orden_2','orden_1'; 'orden_2','orden_0'; 'orden_1','orden_0'};
for mi = 1:2
    if mi == 1, met = 'err'; lab = 'err_demanda'; else, met = 'coste'; lab = 'coste'; end
    fprintf('\n--- %s\n', lab);
    for si = 1:numel(snames)
        for ci = 1:size(cmp,1)
            a = out.(snames{si}).(cmp{ci,1}).(met);
            b = out.(snames{si}).(cmp{ci,2}).(met);
            n = min(numel(a), numel(b));
            if n < 2, continue, end
            d = b(1:n) - a(1:n);
            [~, p, ~, st] = ttest(d);
            star = ''; if p < 0.05, star = ' *'; end
            fprintf('  %-18s%s vs %-10s delta=%+9.4f t=%+7.2f p=%.4f%s\n', ...
                    snames{si}, cmp{ci,1}, cmp{ci,2}, mean(d), st.tstat, p, star);
        end
    end
end
fprintf('\ntiempo: %.1fs\n', toc(t0));
end
